docs/tutorial/writing-tests/test-case.rst

```diff
@@ -0,0 +1,1148 @@
+reason we highly encourage you (and straight up require if you want to
+contribute to the main Checkbox repository), to write unit tests for your
+scripts.
+
+Create a new ``tests/`` directory and a ``test_network_available.py`` file
+inside it.
+
+.. note::
+   You can call your tests however you want but we encourage to make the naming
+   convention uniform at the very least. This tutorial will use the Checkbox
+   naming convention.
+
+The most important thing with your unit tests is that you provide, for each
+function, at least the "happy path" that you have predicted will exist in
+your script. If you have predicted some error path along it (or you have seen
+it happen), create a test for it as well. It is important that each test checks
+for exactly one situation, if possible. Consider the following:
+
+.. code-block:: python
+
+    import unittest
+    import textwrap
+    from unittest import mock
+
+    import network_available
+
+
+    class TestNetworkAvailable(unittest.TestCase):
+
+        @mock.patch("subprocess.check_output")
+        def test_nominal(self, check_output_mock):
+            check_output_mock.return_value = textwrap.dedent(
+                """
+                PING 1.1.1.1 (1.1.1.1) from 192.168.1.100 wlan0: 56(84) bytes
+                64 bytes from 1.1.1.1: icmp_seq=1 ttl=53 time=39.0 ms
+                64 bytes from 1.1.1.1: icmp_seq=2 ttl=53 time=143 ms
+
+                --- 1.1.1.1 ping statistics ---
+                2 packets transmitted, 2 received, 0% packet loss, time 170ms
+                rtt min/avg/max/mdev = 34.980/60.486/142.567/31.077 ms
+                """
+            ).strip()
+            network_available.network_available("wlan0", "90")
+            self.assertTrue(check_output_mock.called)
+
+        @mock.patch("subprocess.check_output")
+        def test_failure(self, check_output_mock):
+            check_output_mock.return_value = textwrap.dedent(
+                """
+                PING 1.1.1.1 (1.1.1.1) from 192.168.1.100 wlan0: 56(84) bytes
+                64 bytes from 1.1.1.1: icmp_seq=1 ttl=53 time=39.0 ms
+
+                --- 1.1.1.1 ping statistics ---
+                10 packets transmitted, a received, 90% packet loss, time 170ms
+                rtt min/avg/max/mdev = 34.980/60.486/142.567/31.077 ms
+                """
+            ).strip()
+            with self.assertRaises(SystemExit):
+                network_available.network_available("wlan0", "0")
+
+.. note::
+   We use ``self.assertTrue(check_output_mock.called)`` instead of
+   ``check_output_mock.assert_called_once()``. The reason is that we have to be
+   compatible (in tests as well!) with python3.5 and
+   ``Mock.assert_called_once`` was introduced in Python3.6. If you don't know
+   when a function was introduced, refer to `the python documentation
+   <https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once>`_.
```

Comment:
```suggestion
   <https://docs.python.org/3/>`_.
```

Reply:
I was linking the topic I was talking about!